File: benchmarks/tsb/bench_format_timedelta_fn.ts
/**
 * Benchmark: formatTimedelta / parseFrac — Timedelta formatting utilities.
 * Mirrors pandas Timedelta string formatting.
 * Outputs JSON: {"function": "format_timedelta_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Timedelta, formatTimedelta, parseFrac, toTimedelta } from "../../src/index.ts";

const WARMUP = 5;
const ITERATIONS = 500;

// Create timedelta instances
const tds = [
  new Timedelta(0),
  new Timedelta(1_000),
  new Timedelta(86_400_000),
  new Timedelta(3_661_001),
  new Timedelta(-7_200_500),
];

for (let i = 0; i < WARMUP; i++) {
  for (const td of tds) {
    formatTimedelta(td);
  }
  parseFrac("123456789");
  parseFrac("000000001");
  toTimedelta(3600, { unit: "s" });
}

const t0 = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  for (const td of tds) {
    formatTimedelta(td);
  }
  parseFrac("123456789");
  parseFrac("000000001");
  toTimedelta(3600, { unit: "s" });
}
const total = performance.now() - t0;

console.log(
  JSON.stringify({
    function: "format_timedelta_fn",
    mean_ms: Math.round((total / ITERATIONS) * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(total * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_format_timedelta_fn.py
"""
Benchmark: pandas Timedelta string formatting — format and parse Timedelta objects.
Mirrors tsb formatTimedelta / parseFrac.
Outputs JSON: {"function": "format_timedelta_fn", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd

WARMUP = 5
ITERATIONS = 500

tds = [
    pd.Timedelta(0),
    pd.Timedelta(seconds=1),
    pd.Timedelta(days=1),
    pd.Timedelta(hours=1, minutes=1, seconds=1, milliseconds=1),
    pd.Timedelta(hours=-2, milliseconds=-500),
]

for _ in range(WARMUP):
    for td in tds:
        str(td)
    pd.Timedelta(seconds=3600)

t0 = time.perf_counter()
for _ in range(ITERATIONS):
    for td in tds:
        str(td)
    pd.Timedelta(seconds=3600)
total = (time.perf_counter() - t0) * 1000

print(json.dumps({
    "function": "format_timedelta_fn",
    "mean_ms": round(total / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total, 3),
}))
